File: cpp/code.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <climits>
#include <fstream>
#include <chrono>

using namespace std;

void dijkstra(unordered_map<char, unordered_map<char, int>> &graph, char start) {
    unordered_map<char, int> distances;
    for (const auto &node : graph) {
        distances[node.first] = INT_MAX;
    }
    distances[start] = 0;

    priority_queue<pair<int, char>, vector<pair<int, char>>, greater<>> pq;
    pq.push({0, start});

    while (!pq.empty()) {
        auto [currentDistance, node] = pq.top();
        pq.pop();

        for (const auto &neighbor : graph[node]) {
            int newDist = currentDistance + neighbor.second;
            if (newDist < distances[neighbor.first]) {
                distances[neighbor.first] = newDist;
                pq.push({newDist, neighbor.first});
            }
        }
    }
}

int main() {
    unordered_map<char, unordered_map<char, int>> graph = {
        {'A', {{'B', 1}, {'C', 4}}},
        {'B', {{'C', 2}, {'D', 5}}},
        {'C', {{'D', 1}}},
        {'D', {}}
    };

    auto start = chrono::high_resolution_clock::now();
    dijkstra(graph, 'A');
    auto end = chrono::high_resolution_clock::now();

    double executionTime = (std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count()) / 1e6;
    ofstream file("result.txt");
    file << "Execution time: " << executionTime << "" << endl;
    file.close();

    return 0;
}

File: javascript/code.js
const dijkstra = (graph, start) => {
    let distances = {};
    let priorityQueue = [[start, 0]];
  
    for (let node in graph) {
      distances[node] = Infinity;
    }
    distances[start] = 0;
  
    while (priorityQueue.length > 0) {
      let [currentNode, currentDist] = priorityQueue.shift();
  
      for (let neighbor in graph[currentNode]) {
        let newDist = currentDist + graph[currentNode][neighbor];
        if (newDist < distances[neighbor]) {
          distances[neighbor] = newDist;
          priorityQueue.push([neighbor, newDist]);
        }
      }
  
      priorityQueue.sort((a, b) => a[1] - b[1]);
    }
  
    return distances;
  };
  
  const graph = {
    A: { B: 1, C: 4 },
    B: { C: 2, D: 5 },
    C: { D: 1 },
    D: {}
  };
  
  // Use process.hrtime for high-resolution timing
  const startTime = process.hrtime();
  dijkstra(graph, "A");
  const endTime = process.hrtime(startTime);
  
  const executionTime = (endTime[0] * 1000) + (endTime[1] / 1e6);
  
  const fs = require("fs");
  fs.writeFileSync("result.txt", `Execution time: ${executionTime.toFixed(6)}\n`);
  